napari/utils/colormaps/colormap.py

```diff
@@ -177,43 +192,25 @@ def map(self, values) -> np.ndarray:
         """
         values = np.atleast_1d(values)
 
-        precast = _cast_labels_to_minimum_dtype_auto(
-            values, len(self.colors) - 1, self.background_value
-        )
-
-        return self._map_precast(precast)
-
-    def _map_precast(self, values) -> np.ndarray:
-        """Map *precast* values to colors.
-
-        When mapping values, we first convert them to a smaller dtype for
-        performance reasons. This conversion changes the label values,
-        even for small labels. This method is used to map values that have
-        already been converted to the smaller dtype.
+        if values.dtype.kind == 'f':
+            values = values.astype(np.int64)
 
-        Parameters
-        ----------
-        values : np.ndarray
-            Values to be mapped. They must have already been downcast using
-            `_cast_labels_to_minimum_dtype_auto`.
-
-        Returns
-        -------
-        np.ndarray of shape (N, M, 4)
-            Mapped colors.
-        """
-        mapped = self.colors[values.astype(np.int64)]
-
-        mapped[values == self.background_value] = 0
-
-        # If using selected, disable all others
-        if self.use_selection:
-            cast_selection = _cast_labels_to_minimum_dtype_auto(
-                np.array([self.selection]),
-                len(self.colors) - 1,
-                self.background_value,
-            )[0]
-            mapped[values != cast_selection] = 0
+        if values.dtype == np.uint8 and "_uint8_colors" in self.__dict__:
+            mapped = self._uint8_colors[values]
+        elif values.dtype == np.uint16 and "_uint16_colors" in self.__dict__:
+            mapped = self._uint16_colors[values]
+        else:
+            background = int(
+                np.array([self.background_value]).astype(values.dtype)[0]
+            )
+            casted_values = _zero_preserving_modulo_naive(
+                values, len(self.colors) - 1, values.dtype, background
+            )
+            mapped = self.colors[casted_values]
+            mapped[casted_values == 0] = 0
+        if self.use_selection and apply_selection:
+            selection2 = np.array([self.selection]).astype(values.dtype)[0]
+            mapped[(values != self.selection) & (values != selection2)] = 0
```

Comment:
I don't actually understand what's happening here since the selection is not getting cast... Is this for situations where there is precision loss? Can you provide an example and maybe add it in a comment?

Reply:
This code is for int8 and int16 data with the negative selection.

Reply:
Yes but I'm confused about why `values` is compared with both `self.selection` and `selection2`. By my understanding, `self.selection` is in the data/raw space and `selection2` is in the texture space. What space are `values` in? Does it make sense to compare them to the selection value in both spaces? And if so, why? 🤔 😕

Reply:
I have removed double comparison and all tests pass. It looks like an artifact from previous iterations (that was required when I did not enforce casting types).